README.md
<p align="center">
 <img src="https://hsto.org/webt/jj/z4/-o/jjz4-ofmx_k18yeslacjnjuzrve.png" width="128" alt="icon">
</p>

# Linter for markdown <sub><sup>| with presets</sup></sub>

![Release][badge_release]
[![Build][badge_ci]][link_actions]
[![Build][badge_build]][link_hub]
[![Docker Pulls][badge_pulls]][link_hub]
[![Issues][badge_issues]][link_issues]
[![License][badge_license]][link_license]

This repository contains source files of docker image (and [github action][github_actions_doc]) for markdown files linting. Additionally we provides setting _(rules)_ for the most useful cases, like changelog file linting.

All docker images always can be found on **[this page][link_hub_tags]**.

## Usage

This image contains [markdownlint-cli][markdownlint-cli] (node-js) and:

- Additional rules for `changelog` file linting _(`/lint/rules/changelog.js`)_
- Configuration file for `changelog` file linting _(it uses additional linting rules, `/lint/config/changelog.yml`)_

> Major image tag can be updated in any time, but all changes will be backwards compatible.

`markdownlint-cli` supports next options (more details can be found [on project page][markdownlint-cli]):

```bash
Usage: markdownlint [options] <files|directories|globs>

Options:
  -h, --help                                  output usage information
  -V, --version                               output the version number
  -f, --fix                                   fix basic errors (does not work with STDIN)
  -s, --stdin                                 read from STDIN (does not work with files)
  -o, --output [outputFile]                   write issues to file (no console)
  -c, --config [configFile]                   configuration file (JSON, JSONC, or YAML)
  -i, --ignore [file|directory|glob]          files to ignore/exclude
  -r, --rules  [file|directory|glob|package]  custom rule files
```

### Environment variables

Some linter execution options can be passed through environment variables _(multiple variables usage are allowed)_:

Environment variable    | Interpretation
----------------------- | --------------
`INPUT_RULES=/foo.js`   | `markdownlint --rules /foo.js ...`
`INPUT_CONFIG=/bar.yml` | `markdownlint --config /bar.yml ...`
`INPUT_FIX=true`        | `markdownlint --fix ...`
`INPUT_OUTPUT=/foo`     | `markdownlint --output /foo ...`
`INPUT_IGNORE=/bar`     | `markdownlint --ignore /bar ...`

### Docker

For example, you can lint your `CHANGELOG.md` file using following command:

```bash
$ docker run --rm \
    -v "$(pwd)/CHANGELOG.md:/CHANGELOG.md:ro" \
    avtodev/markdown-lint:v1 \
    --rules /lint/rules/changelog.js \
    --config /lint/config/changelog.yml \
    /CHANGELOG.md
```

or

```bash
$ docker run --rm \
    -v "$(pwd)/CHANGELOG.md:/CHANGELOG.md:ro" \
    -e INPUT_RULES=/lint/rules/changelog.js \
    -e INPUT_CONFIG=/lint/config/changelog.yml \
    avtodev/markdown-lint:v1 \
    /CHANGELOG.md
```

### GitHub Actions

```yaml
jobs:
  lint-changelog:
    name: Lint changelog file
    runs-on: ubuntu-latest
    steps:
    - name: Check out code
      uses: actions/checkout@v2

    - name: Lint changelog file
      uses: docker://avtodev/markdown-lint:v1 # fastest way
      with:
        rules: '/lint/rules/changelog.js'
        config: '/lint/config/changelog.yml'
        args: './CHANGELOG.md'
        ignore: './one_file.md ./another_file.md' # multiple files must be separated with single space

    # Or using current repository as action:

    - name: Lint changelog file
      uses: avtodev/markdown-lint@v1
      with:
        rules: '/lint/rules/changelog.js'
        config: '/lint/config/changelog.yml'
        args: './CHANGELOG.md'
```

### License

MIT. Use anywhere for your pleasure.

[badge_release]:https://img.shields.io/github/v/release/avto-dev/markdown-lint?include_prereleases&style=flat-square&maxAge=10
[badge_ci]:https://img.shields.io/github/workflow/status/avto-dev/markdown-lint/CI?style=flat-square&maxAge=10
[badge_pulls]:https://img.shields.io/docker/pulls/avtodev/markdown-lint.svg?style=flat-square&maxAge=30
[badge_issues]:https://img.shields.io/github/issues/avto-dev/markdown-lint.svg?style=flat-square&maxAge=30
[badge_build]:https://img.shields.io/docker/cloud/build/avtodev/markdown-lint.svg?style=flat-square&maxAge=30
[badge_license]:https://img.shields.io/github/license/avto-dev/markdown-lint.svg?style=flat-square&maxAge=30
[link_hub]:https://hub.docker.com/r/avtodev/markdown-lint/
[link_hub_tags]:https://hub.docker.com/r/avtodev/markdown-lint/tags
[link_license]:https://github.com/avto-dev/markdown-lint/blob/master/LICENSE
[link_issues]:https://github.com/avto-dev/markdown-lint/issues
[link_actions]:https://github.com/avto-dev/markdown-lint/actions
[markdownlint-cli]:https://github.com/igorshubovych/markdownlint-cli
[github_actions_doc]:https://help.github.com/en/actions


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog][keepachangelog] and this project adheres to [Semantic Versioning][semver].

## v1.5.1 - 2021-09-10

### Fixed

- Docker image name in example part of `README.md`

## v1.5.0 - 2021-01-26

### Changed

- `markdownlint-cli` version updated up to `0.26.0` (from `0.23.2`)

### Fixed

- Multiple files ignoring (you can run current action `with: ignore: 'foo.md bar.md'` with space delimiter for multiple `--ignore` flag appending) [#7]

[#7]:https://github.com/avto-dev/markdown-lint/issues/7

## v1.4.0 - 2020-09-25

### Changed

- Base docker image changed from `node:13-alpine` to `mhart/alpine-node:slim-12` (more lightweight)
- `markdownlint-cli` version updated up to `0.23.2` (from `0.23.1`)
- `markdownlint` packed into single file in docker image

## v1.3.1 - 2020-07-27

### Fixed

- Quoting of `INPUT_*` env variables to allow multiple arguments in 1 env var

## v1.3.0 - 2020-06-19

### Added

- Support for globstar (`**`) by changing the shell to bash [#6]

[#6]:https://github.com/avto-dev/markdown-lint/pull/6

## v1.2.0 - 2020-05-28

### Added

- Support for Keep-a-Changelog's links in the heading [#4]
- Support for using en dashes instead of minus signs [#4]
- Support for passing multiple parameters to `markdown-lint` [#3]

## v1.1.0 - 2020-02-11

### Added

- Changelog rule `CHANGELOG-RULE-004` - Only one "unreleased" version header allowed

### Changed

- Changelog rule `CHANGELOG-RULE-004` enabled in changelog linter config `changelog.yml`

## v1.0.0 - 2020-01-18

### Changed

- First public release

## v0.0.2 - 2020-01-18

### Changed

- GitHub action parameter `path` renamed to `args`

## v0.0.1 - 2020-01-18

### Added

- Docker image auto-building
- Rules set (`CHANGELOG-RULE-001`, `CHANGELOG-RULE-002`, `CHANGELOG-RULE-003`) for changelog file linting
- Config file for changelog file linting

[keepachangelog]:https://keepachangelog.com/en/1.0.0/
[semver]:https://semver.org/spec/v2.0.0.html
[#3]:https://github.com/avto-dev/markdown-lint/pull/3
[#4]:https://github.com/avto-dev/markdown-lint/pull/4


tests/run.sh
#!/usr/bin/env sh
#set -x

runChangelogTest() {
  input_file="$1";          # (string) eg.: "./foo/CHANGELOG.md"
  want_error="${2:-false}"; # (string) eg.: true|false
  error_message="$3";       # (string) eg.: "File not found"

  echo "[TEST] Start test using file [$input_file]";

  if [ ! -f "$input_file" ]; then
    (>&2 echo "[ERROR] Input file $input_file was not found") && exit 50;
  fi;

  output=$(/usr/local/bin/markdownlint \
    --rules /lint/rules/changelog.js \
    --config /lint/config/changelog.yml \
    "$input_file" 2>&1);

  exit_code=$?;

  if [ "$want_error" = "false" ] && [ $exit_code -ne 0 ]; then
    (>&2 echo "$0: [ERROR] Exit code must equals 0. Output [$output]") && exit 50;
  fi;

  if [ "$want_error" = "true" ] && [ $exit_code -eq 0 ]; then
    (>&2 echo "$0: [ERROR] Exit code must be grater then 0") && exit 50;
  fi;

  if [ "$want_error" = "true" ] && [ "$(echo "$output" | grep -c "$error_message")" -eq 0 ]; then
    (>&2 echo "$0: [ERROR] Expected error message [$error_message] was not found in output [$output]") && exit 50;
  fi;

  echo "[OK] Test passed";
}

runChangelogTest ./samples/changelog/correct/sample-1.md false
runChangelogTest ./samples/changelog/correct/sample-2.md false
runChangelogTest ./samples/changelog/correct/sample-3.md false
runChangelogTest ./samples/changelog/incorrect/changes-list-with-ends-punctuation-1.md true "Lists items without punctuation"
runChangelogTest ./samples/changelog/incorrect/changes-list-with-ends-punctuation-2.md true "Lists items without punctuation"
runChangelogTest ./samples/changelog/incorrect/incorrect-cahnges-type-1.md true "Type of changes format"
runChangelogTest ./samples/changelog/incorrect/incorrect-cahnges-type-2.md true "Type of changes format"
runChangelogTest ./samples/changelog/incorrect/missed-first-header.md true "First line in file should be a top"
runChangelogTest ./samples/changelog/incorrect/no-proper-link.md true "Version header format"
runChangelogTest ./samples/changelog/incorrect/wrong-version-header-1.md true "Version header format"
runChangelogTest ./samples/changelog/incorrect/wrong-version-header-2.md true "Version header format"
runChangelogTest ./samples/changelog/incorrect/wrong-version-header-3.md true "Version header format"
runChangelogTest ./samples/changelog/incorrect/duplicated-unreleased-header.md true "Only one 'unreleased' version header"


lint/rules/changelog.js
"use strict";

module.exports = [{
    names: ["CHANGELOG-RULE-001"],
    description: "Version header format",
    tags: ["headings", "headers", "changelog"],
    function: (params, onError) => {
        params.tokens.filter(function filterToken(token) {
            return token.type === "heading_open";
        }).forEach(function forToken(token) {
            if (token.tag === "h2") {
                // eg.: `## v1.0.0`
                if (/^## [vV]?\d+\.\d+\.\d+(-[0-9A-Za-z-.]+|)$/m.test(token.line)) {
                    return;
                }

                // eg.: `## [v1.0.0]`
                if (/^## \[[vV]?\d+\.\d+\.\d+(-[0-9A-Za-z-.]+|)]$/m.test(token.line)) {
                    return;
                }

                // eg.: `## v1.0.0 – 2020-01-01`, `## v1.0.0 - 2020-01-01`
                if (/^## [vV]?\d+\.\d+\.\d+(-[0-9A-Za-z-.]+|) [-–] 20[12][0-9]-[01][0-9]-[0-3][0-9]$/m.test(token.line)) {
                    return;
                }

                // eg.: `## [v1.0.0] – 2020-01-01`, `## [v1.0.0] - 2020-01-01`
                if (/^## \[[vV]?\d+\.\d+\.\d+(-[0-9A-Za-z-.]+|)] [-–] 20[12][0-9]-[01][0-9]-[0-3][0-9]$/m.test(token.line)) {
                    return;
                }

                // eg.: `## unreleased`, `## Unreleased`, `## UNRELEASED`
                if (/^## unreleased$/mi.test(token.line)) {
                    return;
                }

                // eg.: `## [unreleased]`, `## [Unreleased]`, `## [UNRELEASED]`
                if (/^## \[unreleased]$/mi.test(token.line)) {
                    return;
                }

                return onError({
                    lineNumber: token.lineNumber,
                    detail: "Allowed formats: 'vX.X.X(-pre.release)', '[vX.X.X(-pre.release)]', 'vX.X.X(-pre.release) - YYYY-MM-DD', '[vX.X.X(-pre.release)] – YYYY-MM-DD', '[UNRELEASED]', or 'UNRELEASED'",
                    context: token.line
                });
            }
        });
    }
}, {
    names: ["CHANGELOG-RULE-002"],
    description: "Type of changes format",
    tags: ["headings", "headers", "changelog"],
    function: (params, onError) => {
        params.tokens.filter(function filterToken(token) {
            return token.type === "heading_open";
        }).forEach(function forToken(token) {
            if (token.tag === "h3") {
                if (/^### (Added|Changed|Deprecated|Removed|Fixed|Security)$/m.test(token.line)) {
                    return;
                }

                return onError({
                    lineNumber: token.lineNumber,
                    detail: "Allowed types is: Added, Changed, Deprecated, Removed, Fixed or Security",
                    context: token.line
                });
            }
        });
    }
}, {
    names: ["CHANGELOG-RULE-003"],
    description: "Lists items without punctuation at the end",
    tags: ["lists", "changelog"],
    function: (params, onError) => {
        params.tokens.filter(function filterToken(token) {
            return token.type === "list_item_open";
        }).forEach(function forToken(token) {
            if (token.tag === "li") {
                if (/[;,\.]$/m.test(token.line)) {
                    return onError({
                        lineNumber: token.lineNumber,
                        detail: "'.', ';' or ',' at the end of list entry",
                        context: token.line
                    });
                }
            }
        });
    }
}, {
    names: ["CHANGELOG-RULE-004"],
    description: "Only one 'unreleased' version header allowed",
    tags: ["headings", "headers", "changelog"],
    function: (params, onError) => {
        let headers_count = 0;

        params.tokens.filter(function filterToken(token) {
            return token.type === "heading_open";
        }).forEach(function forToken(token) {
            if (token.tag === "h2") {
                if (/^## unreleased$/mi.test(token.line)) {
                    headers_count++;
                }

                if (headers_count >= 2) {
                    return onError({
                        lineNumber: token.lineNumber,
                        detail: "Remove duplicated header",
                        context: token.line
                    });
                }
            }
        });
    }
}];
